### PR TITLE
[ntuple,daos] Adapt I/O backend to DAOSv2 API

### PR DESCRIPTION
**Pull request by glmiotto**
# This Pull request:
Makes changes to ROOT v7 in order to comply with the API for DAOS 2.0.x, affecting the libraries `RPageStorageDaos`, `RDaos` and `libdaos_mock`, as well as the `ntuple_storage_daos` test suite.

## Changes or fixes:
- As of DAOS 2.0, pool service replica ranks are not longer specified in DAOS URIs.  Thus, `daos_pool_connect` now forgoes the `svc_rank_list` argument.
  - Container addresses follow the simplified pattern `daos://($pool)/($container)`. 
  - `RPageStorageDaos` consequently drops parsing enforcement of the previous pattern.
- DAOS 2.0 introduced support for creating pools and containers from a label that is then hashed internally as a UUID. Pool and containers are now identifiable by user-provided labels of up to 127 characters (alphanumeric, colon, period, hyphen or underscore) or by 36-character UUIDs. Labels that match a UUID pattern are explicitly forbidden. 
  - `RDaos` now only supports container creation from user-supplied labels (via `daos_cont_create_with_label`), due to deprecated usage of `daos_cont_create` with user-supplied UUIDs. However, existing containers may still be opened by their UUIDs. 
  - `RPool` and `RContainer` maintain a copy of valid labels (when supplied) and UUIDs (after successfully opening pools and containers respectively)
- Generation of object IDs in `RDaos`: new redundancy, sharding and object type flags; `daos_obj_generate_oid` replaces `daos_obj_generate_id`. 
- The `libdaos_mock` library is updated in lockstep.
- RNTuple backend testing is updated to follow the DAOS 2.0 URI pattern, using labels for default test pool and containers.

## Checklist:

- [x] tested changes locally
- [ ] updated the docs (if necessary)

This PR fixes #

### CODE REVIEW

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (jblomer) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:280]:
```suggestion
   // Verify that it contains only alphanumeric characters or :.-_
```

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (jblomer) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:294]:
```suggestion
   // Check to see if it could be a valid UUID
```

Comment (jblomer) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:294]: Perhaps add a comment why we explicitly check for (and forbid) UUIDs

**Review by jblomer — COMMENTED** [review]
Very nice! A few nits, in particular whole line comments in the code should use the `//` style (which makes it easier to comment blocks including comments with `/* */`).

I leave the final approval up to @jalopezg-r00t

**Comment by glmiotto** [comment]
> Very nice! A few nits, in particular whole line comments in the code should use the `//` style (which makes it easier to comment blocks including comments with `/* */`).
> 
> I leave the final approval up to @jalopezg-r00t

Thanks for the feedback and suggestions, @jblomer! Block comments will be used more judiciously from now on :)

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (glmiotto) [reply tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:294]: This function is taken wholesale from a more recent version of Daos. The check explicitly forbids UUIDs while validating labels because the API so disambiguates IDs for pools or containers - e.g. `daos_cont_create` takes `uuid_t` and `daos_cont_create_with_label` takes `const char*` and generates a UUID internally. I presume the intention is to prevent the user from creating a container whose label matches the UUID of an existing container. 

Perhaps it is excessive to forbid UUID-like labels in our mock library, since it is now entirely label-based and does not hash them into `uuid_t` internally. 

For our `RDaos` backend layer, it would likely be sensible to support both forms of ID rather than switch to labels exclusively, like in this PR. In that case, I could switch back to representing pool and container IDs with `uuid_t` as long as the correct API functions are called at initialization and the hash is retrieved.

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (glmiotto) [reply tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:271]:
```suggestion
```

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (glmiotto) [reply tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:275]:
```suggestion
```

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (glmiotto) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:299]:
```suggestion
      // Implement the check directly to avoid uuid_parse() overhead
```

File: tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx
Comment (glmiotto) [inline tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:404]:
```suggestion
   // | Upper bits contain
   // | OID_FMT_TYPE_BITS (object type)
   // | OID_FMT_META_BITS (object class metadata)
   // | OID_FMT_CLASS_BITS (object class)
   // | 96-bit for upper layer
```

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (glmiotto) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:230]:
```suggestion
   /// Flags to control OC Redundancy
   DAOS_OCH_RDD_DEF = (1 << 0), // Default - use RF prop
   DAOS_OCH_RDD_NO = (1 << 1),  // No redundancy
   DAOS_OCH_RDD_RP = (1 << 2),  // Replication
   DAOS_OCH_RDD_EC = (1 << 3),  // Erasure Code
```

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (glmiotto) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:224]:
```suggestion
/// Flags for oclass hints
```

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (glmiotto) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:161]:
```suggestion
/// DAOS object type
```

File: tree/ntuple/v7/src/RDaos.cxx
Comment (jalopezg-r00t) [inline tree/ntuple/v7/src/RDaos.cxx:87]: We may also explicitly mention a sharding OC hint flag (although as per current DAOS version, it behaves the same if left unspecified).
```suggestion
                            DAOS_OCH_RDD_DEF | DAOS_OCH_SHD_DEF, 0);
```

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (jalopezg-r00t) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:100]: I guess these new declarations are required by the follow-up PR that includes changes in the use of EQs, right?

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (jalopezg-r00t) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:231]: We may need to add `DAOS_OCH_SHD_DEF`, if eventually used in RDaos.cxx.  OTOH, some of the other enumerators can be removed (unused); the less lines we have to maintain here, the better.

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (jalopezg-r00t) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:265]: I would vote for keeping mocks simple.  Therefore, I suggest to get rid of this function; mocks do not really distinguish UUIDs from labels (any string provided by the user is just used as a key in a map).

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (jalopezg-r00t) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:246]: Why is this needed?  Maybe move to the beginning of the file together with other `#include`s.

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (jalopezg-r00t) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:330]: We might even remove this decl (unused now); however, if we want to keep it, see the comment in `libdaos_mock.cxx`

File: tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx
Comment (jalopezg-r00t) [inline tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:286]: Construct a temporary `label_t` (aka `std::string`) instead.  Same for other similar uses below.
```suggestion
   pool->CreateContainer(label_t(reinterpret_cast<char *>(uuid)));
```

File: tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx
Comment (jalopezg-r00t) [inline tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:298]: See comment in `libdaos_mock/daos.h`.

File: tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx
Comment (jalopezg-r00t) [inline tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:169]: We should probably take a reference type instead (same for other uses, e.g. in `CreateContainer()`).
```suggestion
   static RDaosFakePool *GetPool(const label_t &label)
```

**Review by jalopezg-r00t — CHANGES_REQUESTED** [review]
Thanks!  Very nice work, @glmiotto :slightly_smiling_face:!

I have added some comments that we have to address before merging.

**Comment by glmiotto** [comment]
> Thanks! Very nice work, @glmiotto slightly_smiling_face!
> 
> I have added some comments that we have to address before merging.

> Thanks! Very nice work, @glmiotto :slightly_smiling_face:!
> 
> I have added some comments that we have to address before merging.

Great feedback @jalopezg-r00t, thanks as always! I'll get right on it

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (glmiotto) [reply tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:330]: It would make sense to keep this declaration for development purposes if we do support UUID URIs in `RDaos` as well as label ones, even if the mock library and test suites don't use it.

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (glmiotto) [reply tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:100]: Correct, these should have only been added in the next PR. Sorry about the spoilers!

File: tree/ntuple/v7/test/ntuple_storage_daos.cxx
Comment (glmiotto) [reply tree/ntuple/v7/test/ntuple_storage_daos.cxx:13]: Well thought, will do

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (glmiotto) [reply tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:246]: This is due to `isalnum` and `isxdigit` in `daos_label_is_valid` from `daos_prop.h`, which might be removed altogether (or simplified if used for development)

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (glmiotto) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:230]:
```suggestion
   // OC Redundancy flags
   DAOS_OCH_RDD_DEF = (1 << 0), // Default: RF prop
   // OC Sharding flags
   DAOS_OCH_SHD_DEF = (1 << 4), // Default: MAX for array & flat KV, else 1
```

File: tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx
Comment (glmiotto) [inline tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:300]:
```suggestion
   pool->CreateContainer(label_t(label));
```

File: tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx
Comment (glmiotto) [inline tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:314]:
```suggestion
   auto cont = pool->GetContainer(label_t(label));
```

File: tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx
Comment (glmiotto) [inline tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:435]:
```suggestion
   *poh = RDaosHandle::ToHandle(RDaosFakePool::GetPool(label_t(label)));
```

File: tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx
Comment (glmiotto) [inline tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:181]:
```suggestion
   void CreateContainer(const label_t &label)
```

File: tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx
Comment (glmiotto) [inline tree/ntuple/v7/src/libdaos_mock/libdaos_mock.cxx:187]:
```suggestion
   RDaosFakeContainer *GetContainer(const label_t &label)
```

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (glmiotto) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:100]:
```suggestion
```

**Comment by glmiotto** [comment]
@phsft-bot build just on ROOT-ubuntu16/nortcxxmod

**Comment by glmiotto** [comment]
@phsft-bot build

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (jalopezg-r00t) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:272]: In principle, and to keep the header small, we can even simply `return true;` here, as it is not important for the unit tests.

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (jalopezg-r00t) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:284]: This member was only added to silence the `-Wextern-c-compat`, where 0-sized objects are not required to have different addresses.  Thus, we can remove it now. 
```suggestion
```

File: tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h
Comment (jalopezg-r00t) [inline tree/ntuple/v7/inc/ROOT/libdaos_mock/daos.h:299]: Same here.

**Review by jalopezg-r00t — CHANGES_REQUESTED** [review]
Getting closer!  :slightly_smiling_face:

I think this will be the last round of feedback.  After that, we will probably need to squash some commits.

**Review by jalopezg-r00t — APPROVED** [review]
LGTM!

**Pull request merged**
